### PR TITLE
[common] Added new unfused softmax cuda kernel to support causal attention mask

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
In this PR, a new unfused softmax cuda kernel was added to support causal attention mask with arbitrary query sequence length that enables speculative decoding. Max sequence length supported by this and other unfused softmax kernels is 16384. The corresponding test suite is provided in the `test/cpp/operator`.

Some tests fail with very small difference between the reference values, which must be due to the numerical inaccuracy of the softmax.

Performance comparison (duration) of the new kernel vs. unfused upper triangular softmax run on L40 GPU for different sequence lengths:
Performance comparison

### CODE REVIEW

File: tests/cpp/operator/test_causal_softmax.cu [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  using compute_t = float;
+
+  for (int i = 0; i < rows; ++i) {
+    size_t offset = i * cols;
+
+    const int masked_elements = i + cols - rows + 1;
+    compute_t max_value = static_cast<compute_t>(-10'000.f);
+    for (int j = 0; j < masked_elements; ++j) {
+      compute_t tmp = scaling_factor * static_cast<compute_t>(data_in[offset + j]);
+      softmax_out[offset + j] = static_cast<Type>(tmp);
```
Comment (timmoon10) [inline tests/cpp/operator/test_causal_softmax.cu:35]: We are incurring some numerical error by casting the intermediate state from `compute_t` to `Type`. It's not an enormous deal since softmax has good numerical stability, but allocating a temporary `compute_t` buffer would be better.

File: tests/cpp/operator/test_causal_softmax.cu [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+  for (int i = 0; i < rows; ++i) {
+    size_t offset = i * cols;
+
+    const int masked_elements = i + cols - rows + 1;
+    compute_t accumulator = static_cast<compute_t>(0.f);
+    for (int j = 0; j < masked_elements; ++j) {
+      compute_t tmp = static_cast<compute_t>(softmax_in[offset + j])
+                      * static_cast<compute_t>(grad_in[offset + j]);
+      grad_out[offset + j] = static_cast<Type>(tmp);
```
Comment (timmoon10) [inline tests/cpp/operator/test_causal_softmax.cu:78]: We are incurring some numerical error by casting the intermediate state from `compute_t` to `Type`.

**Review by timmoon10 — APPROVED** [review]

File: tests/cpp/operator/test_causal_softmax.cu (hunk 1 above)
Comment (timmoon10) [reply tests/cpp/operator/test_causal_softmax.cu]: The largest test case has sequence length 1024, which doesn't seem that long to me. Do you know which values are outside of the tolerances?

Comment (Oleg-Goncharov) [reply tests/cpp/operator/test_causal_softmax.cu]: I added more test cases which cover key sequence lengths up to 16384 (2^14). Two tests with bfloat16 fail for sequence lengths 256 and 1024:
```
FAILED: (bfloat16, 8, 6, 128, 256, 1.3)
Mismatch at place [1, 2, 18, 88] (266840): 0.00098419189453125 vs 0.000972747802734375

FAILED: (bfloat16, 1, 2, 819, 1024, 2.1)
Mismatch at place [0, 0, 0, 68] (68): 0.00140380859375 vs 0.0013885498046875
```

Comment (Oleg-Goncharov) [reply tests/cpp/operator/test_causal_softmax.cu]: The testing suite uses absolute tolerance `atol=1e-5` for bf16, which is too strict and is beyond the bf16 precision (`2^-7=7.8e-3`), relaxing to `atol=1e-3`

**Review by timmoon10 — APPROVED** [review]
LGTM, pending CI
